### PR TITLE
Support user-supplied histogram in threshold_multiotsu

### LINKED ISSUE #5541: Otsu thresholding admits `hist` as input parameter on `v0.19.0` while multi-Otsu does not
## Description

As it can be seen in the API docs for version `0.19.0.dev0`, multi-Otsu still has the older signature, while 2-class Otsu has the newer one:

https://scikit-image.org/docs/dev/api/skimage.filters.html#skimage.filters.threshold_otsu
https://scikit-image.org/docs/dev/api/skimage.filters.html#skimage.filters.threshold_multiotsu

Feeding the histogram directly has many advantageous use cases, one of them is extracting thresholds for multiple images at the same time (e.g. for video sequences). The feature requested here is to update the signature for multi-Otsu to accept histograms as well.

Currently there is another PR open on multi-Otsu, awaiting integration: https://github.com/scikit-image/scikit-image/pull/5198

I've proposed a workaround there and the author proposed to submit a new, separate PR.
I've opened this issue to introduce and sustain the new PR that I hope to submit in the following days.

## Way to reproduce

This issue is about interfaces and doesn't involve any particular implementation.

## Version information

As mentioned above, the signature inconsistency happens at least in skimage version `0.19.0.dev0`.

### PR DESCRIPTION
**Pull request by andres-fr**
## Description

closes #5541

This is a proposed implementation to close the feature request described in issue https://github.com/scikit-image/scikit-image/issues/5541

For the multi-Otsu implementation, I've mimicked the existing one in the single Otsu thresholding function. I've tested it on my own data and seems to work fine.  I've also updated the function's docstring and added 2 test cases:

1. Check that `ValueError` is raised when neither image nor histogram are given
2. Check that replacing input image with a corresponding `skimage.exposure.histogram` returns the same threshold, while replacing with `np.histogram` returns different ones.

All other thresholding tests seem to pass, and Flake8 seems happy with my additions. I'll be happy to hear any feedback!

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by andres-fr** [comment]
Hi @grlee77, thanks for your time valuable feedback, hopefully I was able to address your comments in the couple of commits I made. Please let me know if any other action is required

**Review by grlee77 — CHANGES_REQUESTED** [review]
Thanks for updating to reuse `_validate_image_histogram`. 

I added suggestions with the requested changes related to f-strings (those had not been addressed yet)

**Comment by andres-fr** [comment]
Sorry I missed that one, I misunderstood your comment. Makes a lot of sense, thanks

**Comment by andres-fr** [comment]
I've approved your suggestions from the gh app and CI seems to be failing (not sure why). I can check it when back home

**Comment by andres-fr** [comment]
I've realized that the suggested fixes were orientative and not working code (parenthesis was missing, didn't parse), I shouldn't have commited them directly. Now it should be fixed, I've added the missing bit, CI seems happy. Sorry for the extra noise!

**Comment by grlee77** [comment]
> I've realized that the suggested fixes were orientative and not working code (parenthesis was missing, didn't parse),

Yes, sorry about that. It was my mistake in copy/pasting the suggestions in the GitHub UI.

**Comment by andres-fr** [comment]
Hi, the UI says that there is 1 change requested, but I'm unable to find it when clicking on it, and elsewhere all suggested changes I found are marked as outdated/resolved. Am I missing something?

**Review by grlee77 — APPROVED** [review]

**Comment by grlee77** [comment]
> Am I missing something?

I think it was just waiting on me to mark my review as "Approved" since I was the one that requested the changes. I fixed one missing space and then did that just now.

File: skimage/filters/thresholding.py
Comment (rfezzani) [inline skimage/filters/thresholding.py:302]:
```suggestion
                image.reshape(-1), nbins, source_range='image', normalize=normalize
```
:wink:

**Comment by grlee77** [comment]
FYI, I updated the title of the PR. I also added "closes #5541" to your original comment at the top of the post. That will tell GitHub to close that issue automatically once this is merged

**Comment by andres-fr** [comment]
Thanks, some of these features are new to me, will keep them in mind. Sorry for the noise

File: skimage/filters/thresholding.py
Comment (andres-fr) [reply skimage/filters/thresholding.py:302]: Hi @rfezzani, I didn't touch things like that one because I wanted to be as least intrussive as possible. Also, ravel is being used in several other places, so changing this would create inconsistencies, and I felt that changing the style of the whole module was out of the scope of this PR. I see that there are [semantic differences between ravel and reshape](https://stackoverflow.com/a/28930580), do you propose to just change this entry or would you suggest changing them all? In the latter case I'd suggest to do so in a separate PR

**Comment by alexdesiqueira** [comment]
Thank you for your work @andres-fr and @grlee77, it looks great!

File: skimage/filters/thresholding.py
Comment (rfezzani) [reply skimage/filters/thresholding.py:302]: @andres-fr, please see the **Notes** section in `np.ravel` [doc](https://numpy.org/doc/stable/reference/generated/numpy.ravel.html) :wink:

Comment (andres-fr) [reply skimage/filters/thresholding.py:302]: Thank you, yes that is aligned with the SO post. I agree with you that `histogram` wouldn't need a copy so we want to provide a flattened view of the array if we can. But I also suppose this is the case for many other uses of `ravel` in the repo. Do you think it would be an idea to open an enhancement issue on this?

Comment (andres-fr) [reply skimage/filters/thresholding.py:302]: OK I dared to `grep -r "ravel" .` in the repo and quite a bit came out. Would `*.c, *.pyx...` files also be in the scope or just `*.py`? Do you know if `ravel/unravel_index` would also be affected by this? (docs don't mention anything)

Comment (rfezzani) [reply skimage/filters/thresholding.py:302]: `*.c`  files are generally generated by Cython, so it doesn't seem in the scope, but `*.pyx` maybe concerned. `ravel/unravel_index` are not concerned by extra/unwanted data copy.
Thank you again for your reactivity, I will open another issue to continue this discussion there :wink:.

Comment (rfezzani) [reply skimage/filters/thresholding.py:302]: please see #5555

**Pull request merged**
